---
title: "Homework2"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/users/n/m/nmphelan/projects/eco_genomics_2025/transcriptomics/mydata")
```

Importing the R libraries

```{r}
library(DESeq2)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(vsn)  
library("pheatmap")
library("vsn")
```

Importing the data

```{r}
# Import the counts matrix
countsTable <- read.table("counts_matrix.txt", header=TRUE, row.names=1)
head(countsTable)
dim(countsTable)

countsTableRound <- round(countsTable) # bc DESeq2 doesn't like decimals (and Salmon outputs data with decimals)
head(countsTableRound)

#import the sample description table
conds <- read.delim("metadata.txt", header=TRUE, stringsAsFactors = TRUE, row.names=1)
head(conds)
```

Running DESeq2 without the interaction term

```{r}
####################################################

### Start working with DESeq2!

####################################################


#Correct column names
colnames(countsTableRound) <- substr(colnames(countsTableRound),
                                     start = 1, stop = 4)
#### Create a DESeq object and define the experimental design here with the tilda
dds <- DESeqDataSetFromMatrix(countData = countsTableRound, colData=conds, 
                              design= ~ generation + line)

dim(dds)

# Filter out genes with too few reads - remove all genes with counts < 15 in more than 75% of 24 samples, so 18)
## suggested by WGCNA on RNAseq FAQ

dds <- dds[rowSums(counts(dds) >= 15) >= 18,]
nrow(dds) 
# How many transcripts have at least 15 reads (a.k.a counts) in 75% of the samples

# Run the DESeq model to test for differential gene expression
dds <- DESeq(dds)

# List the results you've generated
resultsNames(dds)
# Copy the results names 
```
Running DESeq2 with the interaction term

```{r}
#### Create a DESeq object and define the experimental design here with the tilda
dds_interaction <- DESeqDataSetFromMatrix(countData = countsTableRound, colData=conds, 
                              design= ~ generation + line + generation:line)

dim(dds_interaction)

# Filter out genes with too few reads - remove all genes with counts < 15 in more than 75% of 24 samples, so 18)
## suggested by WGCNA on RNAseq FAQ

dds_interaction <- dds_interaction[rowSums(counts(dds) >= 15) >= 18,]
nrow(dds_interaction) 
# How many transcripts have at least 15 reads (a.k.a counts) in 75% of the samples

# Run the DESeq model to test for differential gene expression
dds_interaction <- DESeq(dds_interaction)

# List the results you've generated
resultsNames(dds_interaction)
# Copy the results names 
```
Generating PCAs for data without interaction term

```{r}
# first transform the data for plotting using variance stabilization
vsd <- vst(dds, blind=FALSE)

pcaData <- plotPCA(vsd, intgroup=c("line","generation"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData,"percentVar"))

ggplot(pcaData, aes(PC1, PC2, color=line, shape=generation)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

Another way of visualizing the PCA data
(alpha value and shape is for generation; blue is control & red is treatment)

```{r}
p1=ggplot(pcaData, aes(PC1, PC2)) +
  geom_point(size=5, stroke = 1, aes(fill=line, shape=generation)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  scale_shape_manual(values=c(21,22,23,24), labels = c("G1", "G2","G3", "G4"))+
  #scale_shape_manual(values=c(21,22,23,24) -> circle, square, diamond, triangle
  scale_fill_manual(values=c('#6699CC',"#CC3333"), labels = c("Control", "Treatment"))+
  scale_alpha_manual(values = c(1.0, 0.8, 0.7, 0.3), guide = "none") +
  theme_bw() +
  theme(legend.position = "none") +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 3))+
  theme(text = element_text(size = 20)) +
  theme(legend.title = element_blank())
```

PCAs for DESeq2 with the interaction term

```{r}
# first transform the data for plotting using variance stabilization
vsd <- vst(dds_interaction, blind=FALSE)

pcaData1 <- plotPCA(vsd, intgroup=c("line","generation"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData1,"percentVar"))

ggplot(pcaData1, aes(PC1, PC2, color=line, shape=generation)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

Another way of visualizing the PCA data
(alpha value and shape is for generation; blue is control & red is treatment)

```{r}
p2=ggplot(pcaData1, aes(PC1, PC2)) +
  geom_point(aes(fill = line, shape = generation),
             size = 5, stroke = 1, color = "black", show.legend = TRUE) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  scale_shape_manual(values = c(21, 22, 23, 24), labels = c("G1", "G2", "G3", "G4")) +
  scale_fill_manual(values = c('#6699CC', '#CC3333'), labels = c("Control", "Treatment")) +
  guides(
    fill = guide_legend(override.aes = list(shape = 21, color = "black", alpha = 1))
  ) +
  theme_bw() +
  theme(
    panel.border = element_rect(color = "black", fill = NA, size = 3),
    text = element_text(size = 20),
    legend.title = element_blank()
  )
```

Making the double panel plot and saving it for the write-up

```{r}
Fig1=ggarrange(p1, p2, ncol = 2, nrow = 1, heights = c(1,1), widths = c(1,1.3))

ggsave("Fig1.jpeg", width = 11, height = 4)
```

Testing for differential gene expression using the group model
Staring with the model without interaction between generation and line

```{r}
dds$line <- relevel(dds$line, ref = "control")
dds$generation <- relevel(dds$generation, ref = "G1")


dds$group <- factor(paste0(dds$line, dds$generation))
design(dds) <- ~ group
dds <- DESeq(dds)
resultsNames(dds)

resG1_CvT <- results(dds, contrast=c("group","controlG1","treatmentG1"), alpha = 0.05)

summary(resG1_CvT)
```

Re-order the genes by significance

```{r}
resG1_CvT <- resG1_CvT[order(resG1_CvT$padj),]
head(resG1_CvT)
```

Selected gene TRINITY_DN139369_c0_g5_i1  from the previous output

```{r}
# Counts of specific top interaction gene! (important validation that the normalization, model is working)
d <-plotCounts(dds, gene="TRINITY_DN139369_c0_g5_i1", intgroup = (c("line","generation")), returnData=TRUE)
d

d$line <- factor(d$line, levels = c("control", "treatment"),
                 labels = c("Control", "Treatment"))

p1 <-ggplot(d, aes(x=line, y=count, fill=line, shape=generation)) + 
  theme_bw() + theme(text = element_text(size=20), panel.grid.major=element_line(colour="grey")) +
  scale_shape_manual(values = c(21, 22, 23, 24), labels = c("G1", "G2", "G3", "G4"), name="Generation") +
  scale_fill_manual(
    values = c("Control" = "#6699CC", "Treatment" = "#CC3333"),
    name = "Line"
  ) +
   guides(
    shape = guide_legend(override.aes = list(fill="white", color="black")), 
    fill  = guide_legend(override.aes = list(shape=21, color="black"))      
  ) 
p1 <- p1 + geom_point(mapping = aes(fill = line),
                    position=position_jitter(w=0.2,h=0), size=2, color= "black", show.legend = TRUE, stroke=1)
p1 <- p1 + stat_summary(fun = mean, geom = "point", size=4, fill="#c6abe0", color="black") + labs(x=NULL, y="Counts") + theme(legend.position = "none")+ theme(
  axis.title.y = element_text(margin = margin(r = 7)))
p1

```

Testing for differential gene expression using the group model
Staring with the model with interaction between generation and line

```{r}
dds_interaction$line <- relevel(dds_interaction$line, ref = "control")
dds_interaction$generation <- relevel(dds_interaction$generation, ref = "G1")


dds_interaction$group <- factor(paste0(dds_interaction$line, dds_interaction$generation))
design(dds_interaction) <- ~ group
dds_interaction <- DESeq(dds_interaction)
resultsNames(dds_interaction)

resG1_CvT <- results(dds_interaction, contrast=c("group","controlG1","treatmentG1"), alpha = 0.05)

summary(resG1_CvT)
```

Re-order the genes by significance

```{r}
resG1_CvT <- resG1_CvT[order(resG1_CvT$padj),]
head(resG1_CvT)
```

Selected gene TRINITY_DN139369_c0_g5_i1  from the previous output

```{r}
# Counts of specific top interaction gene! (important validation that the normalization, model is working)
d <-plotCounts(dds_interaction, gene="TRINITY_DN139369_c0_g5_i1", intgroup = (c("line","generation")), returnData=TRUE)
d

d$line <- factor(d$line, levels = c("control", "treatment"),
                 labels = c("Control", "Treatment"))

p2 <-ggplot(d, aes(x=line, y=count, fill=line, shape=generation)) + 
  theme_bw() + theme(text = element_text(size=20), panel.grid.major=element_line(colour="grey")) +
  scale_shape_manual(values = c(21, 22, 23, 24), labels = c("G1", "G2", "G3", "G4"), name="Generation") +
  scale_fill_manual(
    values = c("Control" = "#6699CC", "Treatment" = "#CC3333"),
    name = "Line"
  ) +
   guides(
    shape = guide_legend(override.aes = list(fill="white", color="black")), 
    fill  = guide_legend(override.aes = list(shape=21, color="black"))      
  ) 
p2 <- p2 + geom_point(mapping = aes(fill = line),
                    position=position_jitter(w=0.2,h=0), size=2, color= "black", show.legend = TRUE, stroke=1)
p2 <- p2 + stat_summary(fun = mean, geom = "point", size=4, fill="#c6abe0", color="black") + labs(x=NULL, y="Counts")+ theme(
  axis.title.y = element_text(margin = margin(r = 7)))
p2

```

Trying another gene TRINITY_DN93941_c0_g3_i1 without interaction

```{r}
# Counts of specific top interaction gene! (important validation that the normalization, model is working)
d <-plotCounts(dds, gene="TRINITY_DN93941_c0_g3_i1", intgroup = (c("line","generation")), returnData=TRUE)
d

d$line <- factor(d$line, levels = c("control", "treatment"),
                 labels = c("Control", "Treatment"))

p3 <-ggplot(d, aes(x=line, y=count, fill=line, shape=generation)) + 
  theme_bw() + theme(text = element_text(size=20), panel.grid.major=element_line(colour="grey")) +
  scale_shape_manual(values = c(21, 22, 23, 24), labels = c("G1", "G2", "G3", "G4"), name="Generation") +
  scale_fill_manual(
    values = c("Control" = "#6699CC", "Treatment" = "#CC3333"),
    name = "Line"
  ) +
   guides(
    shape = guide_legend(override.aes = list(fill="white", color="black")), 
    fill  = guide_legend(override.aes = list(shape=21, color="black"))      
  ) 
p3 <- p3 + geom_point(mapping = aes(fill = line),
                    position=position_jitter(w=0.2,h=0), size=2, color= "black", show.legend = TRUE, stroke=1)
p3 <- p3 + stat_summary(fun = mean, geom = "point", size=4, fill="#c6abe0", color="black") + labs(x=NULL, y="Counts")+ theme(legend.position = "none")+ theme(
  axis.title.y = element_text(margin = margin(r = 7)))
p3

```
Adding in the interaction

Trying another gene TRINITY_DN93941_c0_g3_i1 without interaction

```{r}
# Counts of specific top interaction gene! (important validation that the normalization, model is working)
d <-plotCounts(dds_interaction, gene="TRINITY_DN93941_c0_g3_i1", intgroup = (c("line","generation")), returnData=TRUE)
d

d$line <- factor(d$line, levels = c("control", "treatment"),
                 labels = c("Control", "Treatment"))

p4 <-ggplot(d, aes(x=line, y=count, fill=line, shape=generation)) + 
  theme_bw() + theme(text = element_text(size=20), panel.grid.major=element_line(colour="grey")) +
  scale_shape_manual(values = c(21, 22, 23, 24), labels = c("G1", "G2", "G3", "G4"), name="Generation") +
  scale_fill_manual(
    values = c("Control" = "#6699CC", "Treatment" = "#CC3333"),
    name = "Line"
  ) +
   guides(
    shape = guide_legend(override.aes = list(fill="white", color="black")), 
    fill  = guide_legend(override.aes = list(shape=21, color="black"))      
  ) 
p4 <- p4 + geom_point(mapping = aes(fill = line),
                    position=position_jitter(w=0.2,h=0), size=2, color= "black", show.legend = TRUE, stroke=1)
p4 <- p4 + stat_summary(fun = mean, geom = "point", size=4, fill="#c6abe0", color="black") + labs(x=NULL, y="Counts")+ theme(
  axis.title.y = element_text(margin = margin(r = 7)))
p4

```


Making the double panel plot and saving it for the write-up

```{r}
Fig2=ggarrange(p1, p2, p3, p4, ncol = 2, nrow = 2, heights = c(1,1), widths = c(1,1.3))
annotate_figure(Fig2, 
                left = "TRINITY_DN139369_c0_g5_i1                   TRINITY_DN93941_c0_g3_i1")
ggsave("Fig2.jpeg", width = 14, height = 8)
```
